--- create_librimix_from_metadata.py	2023-01-10 17:22:23.141591426 +0800
+++ create_librimix_from_metadata_new.py	2023-01-10 17:52:05.241620084 +0800
@@ -233,13 +233,17 @@
     mixture_id = row['mixture_ID']
     sources_path_list = get_list_from_csv(row, 'source_path', n_src)
     gain_list = get_list_from_csv(row, 'source_gain', n_src)
+    offset = row['offset']
     sources_list = []
     max_length = 0
     # Read the files to make the mixture
-    for sources_path in sources_path_list:
+    for i, sources_path in enumerate(sources_path_list):
         sources_path = os.path.join(librispeech_dir,
                                     sources_path)
         source, _ = sf.read(sources_path, dtype='float32')
+        if i > 0:
+            # padding the second source to simulate an overlap offset
+            source = np.pad(source, (int(offset * 16000), 0), mode="constant")
         # Get max_length
         if max_length < len(source):
             max_length = len(source)
